### PR TITLE
feat: expire stale advanced metrics after not being updated for some time

### PR DESCRIPTION
**Pull request by mmckeen**
# Description

With advanced metrics, high-cardinality labels can cause the metrics export to bloat leading to unbounded memory and resource usage.

This PR sets up an optional TTL for advanced metrics defined in the `MetricsConfiguration` CRD.  By default, the TTL is infinite and cleanup is not tracked or done.

When defined, on a period equal to the TTL metrics which have not been updated within the last TTL duration will be removed from the metrics export.

For counters and gauges that look like counters, this will be treated by Prometheus similar to any other missing metric (for example from an application restart).  As long as functions like `rate` or `increase` are used, calculations will remain accurate.

This also includes a refactor of the base metrics object as an interface, allowing mocking in tests and a clean API contract with child objects that compose it.

## Related Issue

https://github.com/microsoft/retina/issues/1692

## Checklist

- [X] I have read the [contributing documentation](https://retina.sh/docs/Contributing/overview).
- [X] I signed and signed-off the commits (`git commit -S -s ...`). See [this documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.
- [X] I have correctly attributed the author(s) of the code.
- [X] I have tested the changes locally.
- [X] I have followed the project's style guidelines.
- [x] I have updated the documentation, if necessary.
- [x] I have added tests, if applicable.

## Screenshots (if applicable) or Testing Completed

Deployed, modified CRD to enable/disable and change the TTL.

Metrics are re-initialized as expected.

Made sure that CRD validation rejects invalid TTL values.

Also, added comprehensive tests for the TTL-based cleanup logic.

---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

### CODE REVIEW

**Comment by nddq** [comment]
@mmckeen removed the stale tag

File: pkg/module/metrics/basemetricsobject.go [hunk 1]
```diff
@@ -3,25 +3,119 @@
...
 		advEnable:   ctxOptions.IsAdvanced(),
 		ctxOptions:  ctxOptions,
 		l:           fl,
 		contextMode: isLocalContext,
+		lastUpdated: make(map[string]updated),
+		expireFn:    expireOrInfiniteTTL,
+	}
+
+	if expireOrInfiniteTTL != nil {
+		go func() {
```
Comment (nddq) [inline pkg/module/metrics/basemetricsobject.go:163]: I'm concern about the potential goroutine leaks here since there is no proper cancellation. If the `MetricsConfiguration` CRD is reconciled, a new goroutine will spawn and the old one won't get clean up.

File: pkg/module/metrics/basemetricsobject.go [hunk 2]
```diff
@@ -3,25 +3,119 @@
...
+	var bf bytes.Buffer
+
+	for _, l := range lbs {
+		bf.WriteString(l)
+	}
+
+	h := sha256.New()
+	h.Write(bf.Bytes())
+
+	s := string(h.Sum(nil))
```
Comment (nddq) [inline pkg/module/metrics/basemetricsobject.go]: Can't we just concatenate all labels with a unique separator to use it as the key instead? Using sha256 here is kinda overkill imo

File: pkg/module/metrics/basemetricsobject.go [hunk 3]
```diff
@@ -3,25 +3,119 @@
...
+		go func() {
+			for {
+				b.l.Debug(fmt.Sprintf("Expiring metrics: %s", ctxOptions.MetricName))
+				n := b.expire(ttl)
+				b.l.Debug(
+					fmt.Sprintf("Metric expiration finished: %s", ctxOptions.MetricName),
+					zap.Time("next_expiration", time.Now().Add(ttl)),
+					zap.Int("expired", n),
+				)
+				time.Sleep(ttl)
```
Comment (nddq) [inline pkg/module/metrics/basemetricsobject.go]: using `time.Ticker` here might be better here for cleanup purposes

**Review by nddq — CHANGES_REQUESTED** [review]
Sorry for taking a while to respond 😅 Gave this some thoughts

File: pkg/module/metrics/basemetricsobject.go (hunk 1 above)
Comment (mmckeen) [reply pkg/module/metrics/basemetricsobject.go:163]: Fair, let me wrap this in a `sync.Once` or similar

File: deploy/standard/manifests/controller/helm/retina/crds/retina.sh_captures.yaml [hunk 4]
```diff
@@ -165,6 +173,14 @@ spec:
                             type: object
                         type: object
                         x-kubernetes-map-type: atomic
+                      podNames:
```
Comment (mmckeen) [inline deploy/standard/manifests/controller/helm/retina/crds/retina.sh_captures.yaml:176]: Looks like these were missed in previous PRs and this pulled them in when re-generating the CRD manifests, I can manually remove them or put up a second PR to reconcile them.  Let me know 🙇

Comment (nddq) [reply deploy/standard/manifests/controller/helm/retina/crds/retina.sh_captures.yaml:176]: let's put this into a different PR if possible

Comment (mmckeen) [reply deploy/standard/manifests/controller/helm/retina/crds/retina.sh_captures.yaml:176]: Created https://github.com/microsoft/retina/pull/2026

**Comment by mmckeen** [comment]
@nddq appreciate a review on this when you get the chance :)

It's the last PR away from us being able to run the upstream version of Retina rather than our fork.

File: pkg/module/metrics/basemetricsobject.go [hunk 5]
```diff
@@ -3,29 +3,183 @@
...
+
+	ctx, cancel := context.WithCancel(context.Background())
+
 	b := baseMetricObject{
+		RWMutex:     &sync.RWMutex{},
 		advEnable:   ctxOptions.IsAdvanced(),
 		ctxOptions:  ctxOptions,
 		l:           fl,
 		contextMode: isLocalContext,
+		lastUpdated: make(map[string]updated),
```
Comment (nddq) [inline pkg/module/metrics/basemetricsobject.go]: this map will always get allocated even if `ttl < 0`, the same thing with the `ctx` and the mutex, we might be able to refactor this out so that these only get allocated if `ttl > 0` (expiration enabled)

File: pkg/module/metrics/basemetricsobject.go [hunk 6]
```diff
@@ -3,29 +3,183 @@
...
+		ctx:         ctx,
+	}
+
+	if expireOrInfiniteTTL != nil {
+		b.l.Info(
+			"Starting metric expiration routine: "+ctxOptions.MetricName,
+			zap.Duration("ttl", ttl),
+		)
+		go func() {
+			defer cancel()
```
Comment (nddq) [inline pkg/module/metrics/basemetricsobject.go]: nit: this might be unnecessary here since the goroutine will only exit if `<-ctx.Done()` is closed, meaning that `cancel()` is already called.

File: crd/api/v1alpha1/validations/validate_metricconfiguration_test.go
Comment (nddq) [inline crd/api/v1alpha1/validations/validate_metricconfiguration_test.go:1]: nit: add a test for edge cases? 0 or negative values for ttl

File: pkg/module/metrics/dns_test.go [hunk 7]
```diff
@@ -202,114 +204,134 @@ func TestProcessLocalCtx(t *testing.T) {
...
-			tt.d.srcCtx = m
-			tt.d.processLocalCtxFlow(tt.input)
+			ctxOptions := &v1alpha1.MetricsContextOptions{
+				MetricName: utils.DNSResponseCounterName,
+				SourceLabels: []string{
+					podCtxOption,
+					namespaceCtxOption,
+				},
+			}
+			d := NewDNSMetrics(ctxOptions, l, localContext, time.Duration(0))
```
Comment (nddq) [inline pkg/module/metrics/dns_test.go]: i think `time.Duration(0)` can be just 0?

File: pkg/module/metrics/drops_test.go [hunk 8]
```diff
@@ -288,11 +298,40 @@ func TestNewDrop(t *testing.T) {
...
 			f.ProcessFlow(tc.f)
+
+			// There should be no tracked metrics when TTL is infinite
+			assert.Equal(t, 0, len(f.trackedMetricLabels()), "there should be no tracked metrics when TTL is infinite Test Name: %s", tc.name)
+
+			// Test TTL based expiration
+			metricsinit.InitializeMetrics()
+
+			// Set the TTL to something high to ensure that our call to expire is the only one that expires the metrics
+			f = NewDropCountMetrics(tc.opts, log.Logger(), tc.localContext, time.Duration(time.Minute))
```
Comment (nddq) [inline pkg/module/metrics/drops_test.go]: nit: `time.Minute` is already a `Duration` type, so no need to cast it here

**Review by nddq — APPROVED** [review]
lgtm, thanks! Just realized that markdown lint is not required to merge PRs
